### PR TITLE
Feature that allows users to place an order

### PR DESCRIPTION
**Pull request by anamariamilcu**
This PR gives functionality to the 'Order' button and will write the order to the datastore with all the details. OrderPlacingServlet is the one that communicates with the datastore when the user places the order.
The next PR will refactor a little the way Libraries are kept into datastore, to optimize the space used.

### CODE REVIEW

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 1]
```diff
@@ -164,4 +164,33 @@ Entity getClosestLibrary(List<Entity> libraries, Point address) throws ApiExcept
...
+    Query queryUser = new Query("UserInfo")
+        .setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, uid));
+    PreparedQuery resultsUser = datastore.prepare(queryUser);
+    Entity userEntity = resultsUser.asSingleEntity();
+    try {
+      makeOrders(uid, Point.fromAddress((String) userEntity.getProperty("address")),
+          (ArrayList<String>) cartEntity.getProperty("books"));
+    } catch (ApiException | InterruptedException | DataNotFoundException | BadRequestException e) {
+      // Cannot throw exceptions because DoPost only throws IOException.
+      e.printStackTrace();
```
Comment (ak47na) [inline src/main/java/com/google/sps/data/OrderHandler.java]: Why can't you throw all exceptions in `placeOrder` and catch them in `doPost()`?

File: src/main/java/com/google/sps/data/BooksManager.java [hunk 2]
```diff
@@ -30,18 +30,65 @@
...
+    // TODO transaction
+    for (String bookId: bookIds) {
+      Key bookKey = getBookKeyFromDatastore(bookId);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      
+      Query libraryQuery = new Query("LibraryStock").setAncestor(bookKey)
+          .setFilter(new Query.FilterPredicate("libraryId", Query.FilterOperator.EQUAL, (int)library.getLibraryId()));     
+      Entity result = datastore.prepare(libraryQuery).asSingleEntity();
+      // Remove one occurence from library.
+      result.setProperty("stock", ((Number) result.getProperty("stock")).intValue() - 1);
```
Comment (ak47na) [inline src/main/java/com/google/sps/data/BooksManager.java:81]: Where do you check that `stock` and `totalStock` are greater than 0?

Comment (ak47na) [inline src/main/java/com/google/sps/data/BooksManager.java]: I know that this method shouldn't be called for `stock == 0`, but you could add a check for `stock` and `totalStock` to be `> 0`.

**Review by ak47na — APPROVED** [review]

File: src/main/java/com/google/sps/data/BooksManager.java (hunk 2 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/data/BooksManager.java]: Alright, I will add that.

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 1 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/data/OrderHandler.java]: I will fix this.

File: src/main/java/com/google/sps/data/BooksManager.java [hunk 3]
```diff
@@ -30,18 +30,71 @@
...
+    for (String bookId: bookIds) {
+      Key bookKey = getBookKeyFromDatastore(bookId);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      
+      Query libraryQuery = new Query("LibraryStock").setAncestor(bookKey)
+          .setFilter(new Query.FilterPredicate("libraryId", Query.FilterOperator.EQUAL, (int)library.getLibraryId()));     
+      Entity result = datastore.prepare(libraryQuery).asSingleEntity();
+      // Remove one occurence from library.
+      if (((Number) result.getProperty("stock")).intValue() <= 0) {
+        return;
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/BooksManager.java]: How will you know if this succeeded or not?

You need to either return an indicator that this failed (e.g. a boolean) or raise an exception here. Otherwise the caller will expect this to have succeeded, when it didn't.

Please also add tests for what happens when you try to remove books from a library with no stocks. Unit tests are expected to provide coverage for these corner cases.

File: src/test/java/com/google/sps/data/BooksManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    library2.setProperty("stock", 0);
+    datastore.put(library2);
+    Entity library3 = new Entity("LibraryStock", bookEntity.getKey());
+    library3.setProperty("stock", 7);
+    datastore.put(library3);
+    Assert.assertEquals(Arrays.asList(library1, library3), booksManager.getLibrariesForBook("1234"));
+  }
+
+  @Test
+  public void testRemoveBookFromLibrary() throws BadRequestException {
```
Comment (alexcepoi) [inline src/test/java/com/google/sps/data/BooksManagerTest.java]: Maybe also verify that this doesn't remove stock for a book in a different library than the one you requested it from? (could be a separate test, not sure what makes sense).

And also a test on what happens if you try to remove a book from a library which does not have it.

**Review by alexcepoi — CHANGES_REQUESTED** [review]
Looks good, just need to improve your test coverage a bit.

File: src/main/java/com/google/sps/data/BooksManager.java (hunk 3 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/data/BooksManager.java]: I made this method return a boolean, and in the unsuccessful cases it will return false. Then in placeOrder() method from OrderHandler will check if the method was successful or not, if it returns false, it throws a BadRequestException, since you shouldn't actually ever come to the point where you try to remove from a library with no stocks.
Also added tests for when the stock is 0.

File: src/test/java/com/google/sps/data/BooksManagerTest.java (hunk 4 above)
Comment (anamariamilcu) [reply src/test/java/com/google/sps/data/BooksManagerTest.java]: Added those 2 tests.

File: src/main/java/com/google/sps/data/BooksManager.java [hunk 5]
```diff
@@ -30,18 +30,78 @@
...
+        return false;
+      }
+      result.setProperty("stock", ((Number) result.getProperty("stock")).intValue() - 1);
+      datastore.put(result);
+
+      // Remove 1 from total stock.
+      Query bookQuery = new Query("Book").setFilter(new Query.FilterPredicate("bookId", Query.FilterOperator.EQUAL, bookId));
+      Entity bookResult = datastore.prepare(bookQuery).asSingleEntity();
+      if (bookResult.hasProperty("totalStock") == false) {
+        return true;
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/BooksManager.java]: Not sure I understand why was this a success?

This just seems like a database error to me (inconsistent state), so raising an exception might be a better approach.

A comment here would also help.

Also: consider checking both library stock and total stock, running your verifications (both must be > 0), and only then modify the stocks. Otherwise you just modified the library stock, without modifying the total stock, creating further inconsistency in your database. Technically not having inconsistencies is still not guaranteed until you use transactions, but at least it would be less common (e.g. when there's a timeout or connection error, as opposed to all the time).

File: src/main/java/com/google/sps/data/OrderHandler.java [hunk 6]
```diff
@@ -112,14 +112,17 @@ public void addOrderToDatastore(LibraryPoint library, List<String> bookIds, Stri
     for (LibraryPoint library : libraryBookIds.keySet()) {
       ArrayList<String> booksIds = libraryBookIds.get(library);
       addOrderToDatastore(library, booksIds, userId, address);
-      booksManager.removeBooksFromLibrary(library, booksIds);
+      boolean flag = booksManager.removeBooksFromLibrary(library, booksIds);
+      if (flag == false) {
+        throw new BadRequestException("No stock available!");
```
Comment (alexcepoi) [inline src/main/java/com/google/sps/data/OrderHandler.java]: Bad request is generally used when the client request is "syntactically incorrect".
https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletResponse.html

That means the request is invalid regardless of server state.

Maybe DataNotFoundException would be better here?

**Review by alexcepoi — APPROVED** [review]

File: src/main/java/com/google/sps/data/OrderHandler.java (hunk 6 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/data/OrderHandler.java]: Done, it throws DataNotFoundException now

File: src/main/java/com/google/sps/data/BooksManager.java (hunk 5 above)
Comment (anamariamilcu) [reply src/main/java/com/google/sps/data/BooksManager.java]: Changed it.
